pkg/kubelet/rkt/net.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rkt
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	"k8s.io/kubernetes/pkg/kubelet/network/cni"
+	"k8s.io/kubernetes/pkg/kubelet/network/kubenet"
+)
+
+const defaultNetworkName = "default"
+
+type NetworkPluginConfig struct {
+	NetworkPluginName string
+	NetworkPluginDir  string
+	NetworkPlugins    []network.NetworkPlugin
+}
+
+// Init symlinks ${RKT_LOCAL_CONFIG}/net.d and ${RKT_LOCAL_CONFIG}/stage1/net.d
+// to network plugin dir, so that the CNI/kubenet config files can be discoverd by rkt.
+// If the plugin is not CNI or kubenet, that it returns an error.
+func (cfg *NetworkPluginConfig) Init(c *Config) error {
+	switch cfg.NetworkPluginName {
+	case cni.CNIPluginName, kubenet.KubenetPluginName:
+	default:
+		return fmt.Errorf("network plugin %q is not supported", cfg.NetworkPluginName)
+	}
+
+	if cfg.NetworkPluginDir == "" {
+		return nil
+	}
+
+	// Older rkt version supports only ${LOCAL_CONFIG_DIR}/net.d.
+	// Newer version supports both, and will deprecate the ${LOCAL_CONFIG_DIR}/net.d
+	// in the future.
+	// See https://github.com/coreos/rkt/pull/2312#issuecomment-210548103.
+	netdirs := []string{
+		filepath.Join(c.LocalConfigDir, "stage1", "net.d"),
+		filepath.Join(c.LocalConfigDir, "net.d"),
+	}
+
+	succeeded := false
+	for _, dir := range netdirs {
+		if dir == cfg.NetworkPluginDir {
+			// If the --network-plugin-dir is the net.d itself,
+			// do nothing.
+			continue
+		}
+
+		_, err := os.Lstat(dir)
+		if err == nil {
+			if err = os.RemoveAll(dir); err != nil {
```

Comment:
@euank Yeah, good point. But on the other hand I'd like users not to change the value of `--network-plugin-dir` when they switch between runtimes.
One thing we can do is to link the actual config files, but if users put new files under `--network-plugin-dir`, the new files don't get populated, (unless we inotify that dir?)